### PR TITLE
Revamp keras-mxnet docs

### PR DESCRIPTION
**Pull request by sandeep-krishnamurthy**
Revamping Keras-MXNet documentation.

Following are still pending:
1. Tutorial for save_mxnet_model API
2. Clean up benchmark/ docs.

@roywei @kalyc @aaronmarkham

### CODE REVIEW

File: docs/mxnet_backend/installation.md [hunk 1]
```diff
@@ -135,30 +138,25 @@ $  nvidia-smi
...
-    $ export KERAS_BACKEND=mxnet
+```json
+backend: mxnet
+image_data_format: channels_last
 ```
 
-`Option 2` Edit the keras config file
-
-Open "*~/.keras/keras.json*" file and set "*backend=mxnet*".
+We strongly recommend changing the image_data_format to `channels_first`. MXNet is significantly faster on 'channels_first' data. Default is set to 'channels_last' with an objective to be compatible with majority of existing users of Keras. See [performance tips guide](performance_guide.md) for more details.
```
Comment (kalyc) [inline docs/mxnet_backend/installation.md]: are we showing this message prior to model execution as well in the cli?

File: docs/mxnet_backend/performance_guide.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+# Channels First Image Data Format for CNN
+
+When you install the `keras-mxnet`, by default, the following values are set.
+
+```json
+backend: mxnet
+image_data_format: channels_last
+```
+
+We strongly recommend changing the image_data_format to `channels_first`. MXNet is significantly faster on 'channels_first' data. 'channels_first' data format is optimal for training on NVIDIA GPUs with cuDNN, see [here](https://www.tensorflow.org/performance/performance_guide#data_formats) for more details.
```
Comment (aaronmarkham) [inline docs/mxnet_backend/performance_guide.md]: don't link "here"...

File: docs/mxnet_backend/performance_guide.md [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+## Step 2. Update data to channels first format
+
+If your data is in the channels last format (Ex: shape=(256, 256, 3)), you need to convert the data to channels first format (Ex: shape=(3, 256, 256)). 
+
+Keras-MXNet has a utility API [`keras.utils.to_channels_first()`](https://github.com/awslabs/keras-apache-mxnet/blob/master/keras/utils/np_utils.py#L55) that can take a single input or a batch of data in channels last format and return back the data in channels first format.
+
+
+```
+NOTE:
+    Keras dataset API is not consistent across the different dataset. It returns few datasets based on the
```
Comment (aaronmarkham) [inline docs/mxnet_backend/performance_guide.md]: across different datasets. It returns a few datasets...

...and a few datasets in the channels...

File: docs/mxnet_backend/performance_guide.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+
+```
+NOTE:
+    Keras dataset API is not consistent across the different dataset. It returns few datasets based on the 
+    image_data_format in keras.json (Ex: CIFAR10 dataset] and few datasets always in the channels_last format (Ex: 
+    MNIST dataset). It is recommended to always verify the shape of data returned.
+```
+
+## Step 3. Update input_shape argument when constructing the model
```
Comment (aaronmarkham) [inline docs/mxnet_backend/performance_guide.md]: Update the...

File: docs/mxnet_backend/performance_guide.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+You need to update the input shapes in the network to channels first format. For example, using Conv2D on the color images with size 256x256, you need to set:
```
Comment (aaronmarkham) [inline docs/mxnet_backend/performance_guide.md]: 256.

File: docs/mxnet_backend/performance_guide.md [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+For channels last format:
+
+```
+    model.add(Conv2D(32, (3, 3), padding='same', input_shape=(256,256,3)))
+```
+
+For channels first format:
```
Comment (aaronmarkham) [inline docs/mxnet_backend/performance_guide.md]: ...format, set the `input_shape` parameter as follows:

File: docs/mxnet_backend/performance_guide.md [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    Keras dataset API is not consistent across the different dataset. It returns few datasets based on the 
+    image_data_format in keras.json (Ex: CIFAR10 dataset] and few datasets always in the channels_last format (Ex: 
+    MNIST dataset). It is recommended to always verify the shape of data returned.
+```
+
+## Step 3. Update input_shape argument when constructing the model
+
+You need to update the input shapes in the network to channels first format. For example, using Conv2D on the color images with size 256x256, you need to set:
+
+For channels last format:
```
Comment (aaronmarkham) [inline docs/mxnet_backend/performance_guide.md]: ...format, set the `input_shape` parameter as follows:

File: docs/mxnet_backend/performance_guide.md [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+NOTE:
+    It is recommended practice to pass the input_shape based on the shape field of your input tensor. For example:
+    
+    model.add(Conv2D(32, (3, 3), padding='same', input_shape=x_train.shape[1:]))
+
+```
+
+# Optimized MXNet binaries
+
+MXNet PyPi packages are distributed in multiple flavors each optimized for a specific underlying hardware. Please install the correct MXNet flavor optimized for your instance type for the best performance.
```
Comment (aaronmarkham) [inline docs/mxnet_backend/performance_guide.md]: flavors, where each...

File: docs/mxnet_backend/performance_guide.md [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+1. Default: `pip install mxnet`
+ 
+2. Optimized for Intel CPUs with MKL: `pip install mxnet-mkl`
+
+3. Optimized for NVIDIA GPUs with CUDA9: `pip install mxnet-cu90` 
+
+4. Optimized for NVIDIA GPUs and Intel CPUs: `pip install mxnet-cu90-mkl`
+
+5. Refer installation guide for more configurations to build from source: [MXNet Installation guide](https://mxnet.incubator.apache.org/install/index.html)
```
Comment (aaronmarkham) [inline docs/mxnet_backend/performance_guide.md]: Refer to the [MXNet installation guide]()...

File: docs/mxnet_backend/performance_guide.md [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+ 
+2. Optimized for Intel CPUs with MKL: `pip install mxnet-mkl`
+
+3. Optimized for NVIDIA GPUs with CUDA9: `pip install mxnet-cu90` 
+
+4. Optimized for NVIDIA GPUs and Intel CPUs: `pip install mxnet-cu90-mkl`
+
+5. Refer installation guide for more configurations to build from source: [MXNet Installation guide](https://mxnet.incubator.apache.org/install/index.html) 
+
+Please see [MXNet Performance Guide](https://mxnet.incubator.apache.org/faq/perf.html) for more detailed performance tips for MXNet.
```
Comment (aaronmarkham) [inline docs/mxnet_backend/performance_guide.md]: see the MXNet...
delete "for MXNet"

File: docs/mxnet_backend/performance_guide.md [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+
+# Using MXNet Profiler
+
+MXNet has a built-in profiler that provides detailed information about the execution times for each operator. MXNet profiler complements general profiling tools like `nvprof` and `gprof` by summarizing execution time at the operator level, rather than at a function, kernel, or instruction level. Follow [MXNet Performance Guide](https://mxnet.incubator.apache.org/faq/perf.html) for more details on MXNet profiler.
```
Comment (aaronmarkham) [inline docs/mxnet_backend/performance_guide.md]: ...operator. The MXNet profiler...
Follow --> Review the MXNet...
delete "on MXNet profiler"

File: docs/mxnet_backend/using_rnn_with_mxnet_backend.md [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+# Table of Contents
+
+1. [Overview](#overview)
+2. [Variable length inputs are not supported](#variable-length-inputs-are-not-supported)
+3. [Unroll=False is not supported](#unroll=False-is-not-supported)
+4. [Slower CPU training performance](#slower-cpu-training-performance)
+
+# Overview
+
+In this document, we describe the limitations of using RNNs with MXNet backend and workarounds for the same.
```
Comment (aaronmarkham) [inline docs/mxnet_backend/using_rnn_with_mxnet_backend.md]: and available workarounds.

File: docs/mxnet_backend/using_rnn_with_mxnet_backend.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+# Variable length inputs are not supported
+
+MXNet backend does not support variable length inputs in the Recurrent Layers. To overcome this limitation, you can pad the input sequences to prepare fixed length inputs. MXNet backend requires `input_shape` and `unroll=True` parameters while adding SimpleRNN/LSTM/GRU layer.
```
Comment (aaronmarkham) [inline docs/mxnet_backend/using_rnn_with_mxnet_backend.md]: recurrent layers.
The MXNet backend requires both the... and the... adding the

File: docs/mxnet_backend/using_rnn_with_mxnet_backend.md [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+
+```
+NOTE:
+    MXNet does not support symbolic control flow operators. However, this is a work in progress feature. This feature will be soon supported in upcoming releases.
```
Comment (aaronmarkham) [inline docs/mxnet_backend/using_rnn_with_mxnet_backend.md]: ...be supported in an upcoming release.

File: docs/mxnet_backend/using_rnn_with_mxnet_backend.md [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+```
+
+## Transform variable length to fixed length inputs
+
+You should pad the variable length input sequences to make it a fixed length. You can use Keras API - `keras.preprocessing.sequence.pad_sequences` for padding the input.
+ 
+Below we provide an example usecase for the same:
```
Comment (aaronmarkham) [inline docs/mxnet_backend/using_rnn_with_mxnet_backend.md]: Below is an example use case:

File: docs/mxnet_backend/using_rnn_with_mxnet_backend.md [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+```
+NOTE:
+    Padding the input has performance implication due to wasted computation on paddings. You should not pad to 
+    arbitrary large maxlen. It is always optimal to choose maxlen for padding to be equal to the max length of the 
+    input sequences.
+```
+# Unroll=False is not supported
+
+As described above, MXNet backend does not support variable length input and hence `unroll=False` is not supported in RNN layers. Users are expected to provide `input_shape` and set `unroll=True`.
```
Comment (aaronmarkham) [inline docs/mxnet_backend/using_rnn_with_mxnet_backend.md]: Users --> You

File: docs/mxnet_backend/installation.md (hunk 1 above)
Comment (sandeep-krishnamurthy) [reply docs/mxnet_backend/installation.md]: We show this message when user loads keras with mxnet backend.

**Comment by sandeep-krishnamurthy** [comment]
@aaronmarkham - Thanks for details review comments. I have addressed all your review comments.

Added another tutorial - docs/mxnet_backend/save_mxnet_model.md

File: docs/mxnet_backend/save_mxnet_model.md [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+models!
+
+To summarize, all you have to do is to call the `keras.models.save_mxnet_model()` API by passing the trained Keras 
+model to be exported in MXNet model format.
+
+```python
+# ... Assuming you have built and trained a Model
+model.fit(x_train, y_train,
+          batch_size=batch_size,
+          epochs=epochs,
```
[the remark this answers is not in the document]
Comment (sandeep-krishnamurthy) [reply docs/mxnet_backend/save_mxnet_model.md]: Did not understand this.

File: docs/mxnet_backend/save_mxnet_model.md [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+In this section, we show how to load the CNN model, we trained in the previous section, in the MXNet engine. We also 
+perform inference with the native MXNet engine.
+
+```python
+import mxnet as mx
+
+# Step1: Load the model in MXNet
+
+# Use the same prefix and epoch parameters we used in save_mxnet_model API.
+sym, arg_params, aux_params = mx.model.load_checkpoint(prefix='mnist_cnn', epoch=0)
```
[the remark this answers is not in the document]
Comment (sandeep-krishnamurthy) [reply docs/mxnet_backend/save_mxnet_model.md]: No, it is a required parameter for mx.model.load_checkpoint() API

File: docs/mxnet_backend/save_mxnet_model.md [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+test_data = mnist['test_data']
+test_data = test_data.reshape((10000, 1, 784))
+data_iter = mx.io.NDArrayIter(test_data, None, 1)
+result = mod.predict(data_iter)
+idx = 1020
+print("Predicted - ", mx.ndarray.argmax(result[idx]))
+print("Actual - ", labels[idx])
+```
+
+That's it! We trained a CNN model with Keras interface and used MXNet native engine in Python for inference. Also
```
[the remark this answers is not in the document]
Comment (sandeep-krishnamurthy) [reply docs/mxnet_backend/save_mxnet_model.md]: Very nice suggestion, we will try to have a new tutorial for using a pre-trained Keras model(h5py file) for ResNet56-CIFAR10 and export as mxnet model, perform inference in MXNet. Will track this in this Github issue post-release - https://github.com/awslabs/keras-apache-mxnet/issues/84

**Comment by sandeep-krishnamurthy** [comment]
Thanks @aaronmarkham for the feedback. I have addressed your new set of comments.

Pending:
Revamp benchmark results documentation.

**Comment by sandeep-krishnamurthy** [comment]
@aaronmarkham  - Added the final document for your review - benchmark/README.md

**Comment by sandeep-krishnamurthy** [comment]
This PR is now complete with all docs for mxnet backend rewritten/revamped. 
Pending:
1. Review from @aaronmarkham for benchmark/README.md file.
2. Updating latest benchmark results in benchmark/README.md by @roywei 

@lupesko - Any other comments?

File: docs/mxnet_backend/save_mxnet_model.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (aaronmarkham) [reply docs/mxnet_backend/save_mxnet_model.md]: But you're using epoch=0, when we just did training in the previous section. Should it not match the amount of training done? Or should it say that epoch doesn't really matter so you can just pass 0?

File: benchmark/benchmark_result/RNN_result.md [hunk 20]
```diff
@@ -1,3 +1,40 @@
...
+|---|---|---|---|---|
+|  C5.18X Large | 0  | 128  | 24 sec, N/A | 14 sec, N/A|
+|  P3.8X Large |  1 |  128 | 13 sec, 792 MB | 12 sec, 15360 MB|
+|  P3.8X Large | 4  | 128  | 12 sec, 770 MB | 21 sec, 15410 MB |
+|  P3.16X Large | 8  | 128  | TBD | TBD |
+
+
 # Detailed RNN Benchmark Results
 
 Below is the result of GPU memory usage while running LSTM model on Synthetic, Nietzsche, and WikiText-2 character level dataset.
```
Comment (aaronmarkham) [inline benchmark/benchmark_result/RNN_result.md]: This is left for the reader's interpretation, and I'm not sure what to think about it. No commentary to assist?
